pkg/forwarder/forwarder.go

```diff
@@ -264,6 +271,37 @@ func (f *DefaultForwarder) Stop() {
 	log.Info("DefaultForwarder stopped")
 }
 
+func (f *DefaultForwarder) healthCheckLoop() {
+	log.Debug("Waiting for APIkey validity to be confirmed.")
+	// Wait until we confirmed we have one valid APIkey to become healthy
+	waitTicker := time.NewTicker(time.Second)
+	validKey := false
+	for range waitTicker.C {
+		apiKeyStatus.Do(func(entry expvar.KeyValue) {
+			if entry.Value == &apiKeyValid {
+				validKey = true
+			}
+		})
+		if validKey {
+			break
+		}
+	}
+	waitTicker.Stop()
+
+	log.Debug("APIkey is valid, forwarder is healthy.")
+
+	// Once we're healthy, make sure we don't drop packets
+	for range f.health.C {
+		// By ranging, we will exit the goroutine when the channel closes.
+		// If we dropped a transaction (full queue or invalid APIkey),
+		// we exit the loop and let the forwarder become unhealthy.
+		if transactionsExpvar.Get("Dropped") != nil {
```

Comment:
I think this condition is too strict: a forwarder can start dropping transactions while still being completely healthy, for instance if there's a network issue that lasts for a significant amount of time and makes the forwarder fill up the retry queue. 

If I understand correctly, an agent with an unhealthy component would typically be restarted. In that case, if transactions are dropped because of a lasting network issue, restarting the Agent would be counter-productive: it would make the Agent lose _all_ transactions, while not helping with the root cause (the network issue).

I'm not sure what would be a good way to detect deadlocked/blocked workers, but overall I think it's more desirable to avoid false positives of unhealthiness than false negatives.

WDYT? (and let me know if my interpretation is wrong :) )

Reply:
I think a full `highPrio` queue should set the forwarder as unhealthy, as it's unable to process incoming transactions. This is the logic added here: https://github.com/DataDog/datadog-agent/blob/master/pkg/forwarder/forwarder.go#L307

The issue is that indeed `Dropped` is incremented elsewhere (on invalid APIkey full retry queue..._, shall this condition also increment a new `IntakeDropped` that we'll use for the healthcheck?